### PR TITLE
Let Kotlin DSL script compilation not fix JVM target to 8

### PR DESCRIPTION
**Pull request by eskatos**
Previously, the compilation of `.gradle.kts` scripts always used Java 8 as the Kotlin JVM target.
Starting with this PR, it now uses the version of the JVM running the build.

If your team is using e.g. Java 11 to run Gradle, this allows you to use Java 11 librairies and language features in your build scripts.

----

Implementation notes

* [x] IDE editor uses the Gradle build JVM in its context thus reflects the JVM target used to compile the scripts
* [x] Generated Gradle API Kotlin DSL extensions JAR is still compiled with target JVM 8
* [x] Selected JVM target is part of the cache key for script compilation
* [x] Accessors generation still uses JVM 8
    JVM types in accessors are not accessible anyway
    https://github.com/gradle/gradle/issues/22968
    Adding JVM selection would not provide any enhancement, would be convoluted to test and add complexity at this point

Related optimistic fixes

* [x] `plugins {}` block compilation for precompiled scripts now honors configured `jvmTarget`

----

* Fixes https://github.com/gradle/gradle/issues/21041

### CODE REVIEW

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/jvm/JavaClassUtil.groovy [hunk 1]
```diff
@@ -45,5 +46,12 @@ class JavaClassUtil {
         return javaClass.name.replace('.', '/') + '.class'
     }
 
+    static int getClassMajorVersion(JavaVersion javaVersion) {
+        if (javaVersion == JavaVersion.VERSION_HIGHER) {
+            throw new UnsupportedOperationException("Unable to provide class file major version for '$javaVersion'")
+        }
+        return javaVersion.majorVersion.toInteger() + 44
```
[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/jvm/JavaClassUtil.groovy:53]: This is correct for all the existing java versions. How would it be better to have the hardcoded mapping for existing versions?

[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/jvm/JavaClassUtil.groovy]: Good point, done

**Review by ljacomet — APPROVED** [review]
Looks good overall, one comment about commented code.

**Comment by eskatos** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
Sorry, need to merge my PR first to unblock master. https://github.com/gradle/gradle/pull/22974

@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Comment by Amejonah1200** [comment]
I think one major flaw of this PR was that if you run Gradle with JVM 25 but the bundled Kotlin can only do JVM 24, it will fail.

Is my assumption correct that this PR is at fault that newer JVM versions cannot be used to run Gradle because bundled kotlin compiler for the scripts cannot output the newer versin?

**Comment by eskatos** [comment]
@Amejonah1200, we actually fallback to the highest JVM target that the Kotlin compiler supports automatically. This PR is not at fault.

**Pull request merged**